### ISSUE TITLE
Roadmap for features

### ISSUE THREAD

**Issue by daliwali** [samselikoff/ember-cli-mirage#938]
In issue https://github.com/emberjs/guides/issues/1695#issuecomment-253596178 @samselikoff commented:

>We have an issue on the backlog to make Mirage able to serve out of Ember CLI (from node), which would bring back the network tab. The next goal after that is to have Mirage read model definitions from Ember Data, if you happen to be using it.

I could not find related issues here. Also, are there are plans for:

- Built-in support for more optional JSON API features, such as sparse fieldsets, pagination, sorting, filtering, etc.
- Database features such as persistence, referential integrity, automatic inverse updates, etc.

**Comment by samselikoff** [comment]
You can follow the ORM project ([projects](https://github.com/samselikoff/ember-cli-mirage/projects)) to see what needs to happen before we can implement the auto-schema generation from your Ember Data models. This can be worked on in tandem with the project that will serve Mirage from node.

After I finish the ORM work I am going to break these projects down and recruit project leads.

Additional JSON API features are welcome, I can also delimit the work that needs to happen here.

If we end up adding a feature to persist the database, it would be something lightweight, perhaps a simple file or SQLite. Mirage is a developer's tool and it is not designed to be used as a production server. Once Mirage can serve from Node it will have persistent state for the duration of a "session" (an ember app served with `ember s`), which will address many needs.

I'm not sure what you mean by automatic inverse updates, the database should be consistent (this is the ORMs responsibility). There is some referential integrity built in to the ORM, but that's all at the moment.

**Comment by acorncom** [comment]
@sapeien Are your questions above in relation to the in-house add-on you mentioned here (https://github.com/emberjs/guides/issues/1695#issuecomment-255694723)?

> Mirage is a developer's tool and it is not designed to be used as a production server.

@samselikoff regarding the above, could you see a way for Mirage to be the initial building block toward building out a production server?  Part of my interest in helping introduce folks to Mirage would be to see if through using Mirage they could then further iterate into a fully-fleshed out Node backend (potentially something like https://github.com/denali-js/denali which @davewasmer has been the driving force behind but which I've been helping kick the tires on).

Is there someway we might be able to put all these creative juices together (along with Erik Bryn's interest in "blessing" a Node framework for the Ember ecosystem) to end up channeling this energy into building an opinionated Node framework that we'd all like to use?

**Comment by sapeien** [comment]
@acorncom,

>Is there someway we might be able to put all these creative juices together (along with Erik Bryn's interest in "blessing" a Node framework for the Ember ecosystem) to end up channeling this energy into building an opinionated Node framework that we'd all like to use?

That would be ideal, really. One should take a look at this list to see how much fragmentation there already is: http://jsonapi.org/implementations/#server-libraries-node-js

Of these implementations, I would say that almost all of them are incomplete (only implementing just a few JSON API features) or unmaintained. There are even more which are underway: https://github.com/starspot/starspot

The reason why Fortune.js is first on that list is that it's the first attempt at creating a back-end that can serialize all of JSON API, way back in 2013. Since 2015 it has been rewritten so that JSON API is just one of many serialization formats. It has the most complete implementation of JSON API in Node.js that I know of, and the test suite I wrote covers all of the optional features of the spec. From my perspective, trying to implement all of this in Mirage would be duplicating work that I and many others have done already.

Regarding the addon I wrote, it's just 150 lines of code and small enough to just publish as a gist, what it does is mostly static analysis of Ember Data models. I am not really involved in the Ember community and at that time there was no Ember CLI, so I would have released something in 2014 if I were using Ember at the time. We are hesitant to release it because we at @simplabs are worried that now it would compete with Mirage, basically political reasons.
